# Demographic Data Analyzer

#### The second project of the "[*Data Analysis with Python*](https://www.freecodecamp.org/learn/data-analysis-with-python/)" Certification.

#### A [freeCodeCamp.org](https://www.freecodecamp.org/) course

![purple-divider](https://user-images.githubusercontent.com/7065401/52071927-c1cd7100-2562-11e9-908a-dde91ba14e59.png)

### The Goal

In this challenge, we must analyze demographic data. We are given a dataset of demographic data that was extracted from the 1994 US Census database.

We must use Pandas to answer some questions, rounding all decimals to the nearest tenth.

### Hands on!

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('adult.data.csv')

In [3]:
df.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

### 1. How many people of each race are represented in this dataset?

This should be a Pandas series with race names as the index labels. (`race` column)

In [4]:
race_count = df['race'].value_counts()
race_count

White                 27816
Black                  3124
Asian-Pac-Islander     1039
Amer-Indian-Eskimo      311
Other                   271
Name: race, dtype: int64

Is this a Pandas series?

In [5]:
isinstance(race_count, pd.Series)

True

### 2. What is the average age of men?

In [6]:
average_age_men = df[df['sex'] == 'Male']['age'].mean()
average_age_men = round(average_age_men, 1)
average_age_men

39.4

### 3. What is the percentage of people who have a Bachelor's degree?

In [7]:
percentage_bachelors = df[df['education'] == 'Bachelors'].shape[0]/df.shape[0]*100
percentage_bachelors = round(percentage_bachelors, 1)
percentage_bachelors

16.4

### 4. What percentage of people with advanced education (`Bachelors`, `Masters`, or `Doctorate`) make more than 50K?

In [8]:
higher_education = df[df['education'].isin(['Bachelors', 'Masters', 'Doctorate'])].shape[0]

In [9]:
higher_education_rich = df[(df['education'].isin(['Bachelors', 'Masters', 'Doctorate'])) & (df['salary']=='>50K')].shape[0]/higher_education*100
higher_education_rich = round(higher_education_rich, 1)
higher_education_rich

46.5

### 5. What percentage of people without advanced education make more than 50K?

In [10]:
lower_education = df[~df['education'].isin(['Bachelors', 'Masters', 'Doctorate'])].shape[0]

In [11]:
lower_education_rich = df[~(df['education'].isin(['Bachelors', 'Masters', 'Doctorate'])) & (df['salary']=='>50K')].shape[0]/lower_education*100
lower_education_rich = round(lower_education_rich, 1)
lower_education_rich

17.4

### 6. What is the minimum number of hours a person works per week?

In [12]:
min_work_hours = df['hours-per-week'].min()
min_work_hours

1

### 7. What percentage of the people who work the minimum number of hours per week have a salary of more than 50K?

In [13]:
num_min_workers = df[df['hours-per-week'] == min_work_hours].shape[0]

In [14]:
rich_percentage = df[(df['hours-per-week'] == min_work_hours) & (df['salary']=='>50K')].shape[0] / num_min_workers *100
rich_percentage = round(rich_percentage, 1)
rich_percentage

10.0

### 8. What country has the highest percentage of people that earn more than 50K and what is that percentage?

In [15]:
native_country_list = df['native-country'].value_counts()
native_country_list

United-States                 29170
Mexico                          643
?                               583
Philippines                     198
Germany                         137
Canada                          121
Puerto-Rico                     114
El-Salvador                     106
India                           100
Cuba                             95
England                          90
Jamaica                          81
South                            80
China                            75
Italy                            73
Dominican-Republic               70
Vietnam                          67
Guatemala                        64
Japan                            62
Poland                           60
Columbia                         59
Taiwan                           51
Haiti                            44
Iran                             43
Portugal                         37
Nicaragua                        34
Peru                             31
France                      

We note that we have `583` citizens of whom we do not know their nationality of origin. At the moment this is not a problem.

Let's calculate how many people from each country earn more than 50K. We will create a list with this information:

In [16]:
aux = []

for country in native_country_list.index:
    aux.append(df[(df['native-country'] == country) & (df['salary']=='>50K')].shape[0])

Now we organize this information in a new Data Frame:

In [17]:
new_df = pd.DataFrame(native_country_list)
new_df['salary-more-50K'] = aux
new_df

native-country  salary-more-50K
United-States                        29170             7171
Mexico                                 643               33
?                                      583              146
Philippines                            198               61
Germany                                137               44
Canada                                 121               39
Puerto-Rico                            114               12
El-Salvador                            106                9
India                                  100               40
Cuba                                    95               25
England                                 90               30
Jamaica                                 81               10
South                                   80               16
China                                   75               20
Italy                                   73               25
Dominican-Republic                      70                2
Vietnam                                 67                5
Guatemala                               64                3
Japan                                   62               24
Poland                                  60               12
Columbia                                59                2
Taiwan                                  51               20
Haiti                                   44                4
Iran                                    43               18
Portugal                                37                4
Nicaragua                               34                2
Peru                                    31                2
France                                  29               12
Greece                                  29                8
Ecuador                                 28                4
Ireland                                 24                5
Hong                                    20                6
Cambodia                                19                7
Trinadad&Tobago                         19                2
Laos                                    18                2
Thailand                                18                3
Yugoslavia                              16                6
Outlying-US(Guam-USVI-etc)              14                0
Honduras                                13                1
Hungary                                 13                3
Scotland                                12                3
Holand-Netherlands                       1                0

And finally we add a new column to answer the questions.

In [18]:
new_df['percentage-rich'] = new_df['salary-more-50K'] / new_df['native-country'] *100
new_df['percentage-rich'] = new_df['percentage-rich'].round(decimals = 1)
new_df

native-country  salary-more-50K  percentage-rich
United-States                        29170             7171             24.6
Mexico                                 643               33              5.1
?                                      583              146             25.0
Philippines                            198               61             30.8
Germany                                137               44             32.1
Canada                                 121               39             32.2
Puerto-Rico                            114               12             10.5
El-Salvador                            106                9              8.5
India                                  100               40             40.0
Cuba                                    95               25             26.3
England                                 90               30             33.3
Jamaica                                 81               10             12.3
South                                   80               16             20.0
China                                   75               20             26.7
Italy                                   73               25             34.2
Dominican-Republic                      70                2              2.9
Vietnam                                 67                5              7.5
Guatemala                               64                3              4.7
Japan                                   62               24             38.7
Poland                                  60               12             20.0
Columbia                                59                2              3.4
Taiwan                                  51               20             39.2
Haiti                                   44                4              9.1
Iran                                    43               18             41.9
Portugal                                37                4             10.8
Nicaragua                               34                2              5.9
Peru                                    31                2              6.5
France                                  29               12             41.4
Greece                                  29                8             27.6
Ecuador                                 28                4             14.3
Ireland                                 24                5             20.8
Hong                                    20                6             30.0
Cambodia                                19                7             36.8
Trinadad&Tobago                         19                2             10.5
Laos                                    18                2             11.1
Thailand                                18                3             16.7
Yugoslavia                              16                6             37.5
Outlying-US(Guam-USVI-etc)              14                0              0.0
Honduras                                13                1              7.7
Hungary                                 13                3             23.1
Scotland                                12                3             25.0
Holand-Netherlands                       1                0              0.0

What country has the highest percentage of people that earn more than 50K?

In [19]:
highest_earning_country = new_df['percentage-rich'].idxmax()
highest_earning_country

'Iran'

What is that percentage?

In [20]:
highest_earning_country_percentage = new_df.loc[highest_earning_country]['percentage-rich']
highest_earning_country_percentage

41.9

### 9. Identify the most popular occupation for those who earn more than 50K in India.

In [21]:
IN_richest = df[(df['native-country'] == 'India') & (df['salary']=='>50K')]['occupation'].value_counts()
IN_richest

Prof-specialty      25
Exec-managerial      8
Other-service        2
Tech-support         2
Transport-moving     1
Sales                1
Adm-clerical         1
Name: occupation, dtype: int64

In [22]:
top_IN_occupation = IN_richest.idxmax()
top_IN_occupation

'Prof-specialty'

### Appendix

Now, we can submit the project using the following function:

In [23]:
def calculate_demographic_data(print_data=True):
    # Read data from file
    df = pd.read_csv('adult.data.csv')

    # How many of each race are represented in this dataset? This should be a Pandas series with race names as the index labels.
    race_count = df['race'].value_counts()

    # What is the average age of men?
    average_age_men = df[df['sex'] == 'Male']['age'].mean()
    average_age_men = round(average_age_men, 1)

    # What is the percentage of people who have a Bachelor's degree?
    percentage_bachelors = df[df['education'] == 'Bachelors'].shape[0]/df.shape[0]*100
    percentage_bachelors = round(percentage_bachelors, 1)

    # What percentage of people with advanced education (`Bachelors`, `Masters`, or `Doctorate`) make more than 50K?
    # What percentage of people without advanced education make more than 50K?

    # with and without `Bachelors`, `Masters`, or `Doctorate`
    higher_education = df[df['education'].isin(['Bachelors', 'Masters', 'Doctorate'])].shape[0]
    ## alternative: higher_education = df[(df['education'] == 'Bachelors') | (df['education'] =='Masters') | (df['education'] =='Doctorate')].shape[0]
    
    lower_education = df[~df['education'].isin(['Bachelors', 'Masters', 'Doctorate'])].shape[0]
    
    
    # percentage with salary >50K
    higher_education_rich = df[(df['education'].isin(['Bachelors', 'Masters', 'Doctorate'])) & (df['salary']=='>50K')].shape[0]/higher_education*100
    higher_education_rich = round(higher_education_rich, 1)
    
    lower_education_rich = df[~(df['education'].isin(['Bachelors', 'Masters', 'Doctorate'])) & (df['salary']=='>50K')].shape[0]/lower_education*100
    lower_education_rich = round(lower_education_rich, 1)
    
    # What is the minimum number of hours a person works per week (hours-per-week feature)?
    min_work_hours = df['hours-per-week'].min()

    # What percentage of the people who work the minimum number of hours per week have a salary of >50K?
    num_min_workers = df[df['hours-per-week'] == min_work_hours].shape[0]

    rich_percentage = df[(df['hours-per-week'] == min_work_hours) & (df['salary']=='>50K')].shape[0] / num_min_workers *100
    rich_percentage = round(rich_percentage, 1)
    
    # What country has the highest percentage of people that earn >50K?
    
    native_country_list = df['native-country'].value_counts()
    
    aux = []
    for country in native_country_list.index:
        aux.append(df[(df['native-country'] == country) & (df['salary']=='>50K')].shape[0])
        
    new_df = pd.DataFrame(native_country_list)
    new_df['salary-more-50K'] = aux
    new_df['percentage-rich'] = new_df['salary-more-50K'] / new_df['native-country'] *100
    new_df['percentage-rich'] = new_df['percentage-rich'].round(decimals = 1)
        
    highest_earning_country = new_df['percentage-rich'].idxmax()
    highest_earning_country_percentage = new_df.loc[highest_earning_country]['percentage-rich']

    # Identify the most popular occupation for those who earn >50K in India.
    
    IN_richest = df[(df['native-country'] == 'India') & (df['salary']=='>50K')]['occupation'].value_counts()
    top_IN_occupation = IN_richest.idxmax()

    # DO NOT MODIFY BELOW THIS LINE

    if print_data:
        print("Number of each race:\n", race_count) 
        print("Average age of men:", average_age_men)
        print(f"Percentage with Bachelors degrees: {percentage_bachelors}%")
        print(f"Percentage with higher education that earn >50K: {higher_education_rich}%")
        print(f"Percentage without higher education that earn >50K: {lower_education_rich}%")
        print(f"Min work time: {min_work_hours} hours/week")
        print(f"Percentage of rich among those who work fewest hours: {rich_percentage}%")
        print("Country with highest percentage of rich:", highest_earning_country)
        print(f"Highest percentage of rich people in country: {highest_earning_country_percentage}%")
        print("Top occupations in India:", top_IN_occupation)

    return {
        'race_count': race_count,
        'average_age_men': average_age_men,
        'percentage_bachelors': percentage_bachelors,
        'higher_education_rich': higher_education_rich,
        'lower_education_rich': lower_education_rich,
        'min_work_hours': min_work_hours,
        'rich_percentage': rich_percentage,
        'highest_earning_country': highest_earning_country,
        'highest_earning_country_percentage': highest_earning_country_percentage,
        'top_IN_occupation': top_IN_occupation
    }